### PR TITLE
UCP: API - adding priority attribute to ucp requests

### PR DESCRIPTION
**Pull request by roiedanino**
## What
Adding priority attribute to `ucp_request_param_t`

## Why ?
Upcoming feature - high-priority messages, on supporting transports.

### CODE REVIEW

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -734,12 +734,13 @@ typedef enum {
...
                                                         operation cannot be
                                                         completed immediately */
-    UCP_OP_ATTR_FLAG_MULTI_SEND     = UCS_BIT(19)  /**< optimize for bandwidth of
+    UCP_OP_ATTR_FLAG_MULTI_SEND     = UCS_BIT(19),  /**< optimize for bandwidth of
                                                         multiple in-flight operations,
                                                         rather than for the latency
                                                         of a single operation.
                                                         This flag and UCP_OP_ATTR_FLAG_FAST_CMPL
                                                         are mutually exclusive. */
+    UCP_OP_ATTR_FIELD_PRIORITY      = UCS_BIT(20)   /**< priority field*/
```
Comment (brminich) [inline src/ucp/api/ucp.h:743]:
```suggestion
    UCP_OP_ATTR_FIELD_PRIORITY      = UCS_BIT(20)   /**< priority field */
```

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -815,6 +816,19 @@ enum ucp_am_handler_param_field {
...
+ * @ingroup UCP_COMM
+ * @brief UCP request priority level
+ * 
+ * The enumeration lists available priority levels which are set per request, 
+ * Message priority is an optional attribute of @ref ucp_request_param_t, 
+ * value will be propogated to transports which are capable of prioritizing messages.
+*/
+typedef enum {
+    UCP_PRIORITY_DEFAULT = UCS_BIT(0),
+    UCP_PRIORITY_HIGH    = UCS_BIT(1)
```
Comment (brminich) [inline src/ucp/api/ucp.h:829]: why is it a flag? I think we can't combine different prio levels for one request

Comment (roiedanino) [reply src/ucp/api/ucp.h]: Agree, my bad

**Review by brminich — DISMISSED** [review]

**Comment by shamisp** [comment]
what is the use case and how it is going to be implemented ? I don't just stating the layers is sufficient for QoS definition.

**Comment by roiedanino** [comment]
> what is the use case and how it is going to be implemented ? I don't thinking just stating "the priorities" is sufficient for QoS definition.

### Use Case
Prioritizing specific messages when possible, UCX will rely on HW configuration, for example:
HW will be configured with a high-priority SL and UCX will be configured accordingly: 
UCX_IB_HIGH_PRIORITY_SL=<hw configured sl>

### Implementation:
1. Some of the `ifaces` will have priority messages capability
2. Each `uct_ep` created at UCP level will have a matching high-priority `uct_ep` if iface is capable 
3. For each request: `request_param_t` will indicate the priority level: 
in case the selected lane / the ep supports prioritizing, the priority `uct_ep` will be used, in `ib` that ep will be configured to use the UCX_IB_HIGH_PRIORITY_SL value as sl.

**Comment by shamisp** [comment]
> > what is the use case and how it is going to be implemented ? I don't thinking just stating "the priorities" is sufficient for QoS definition.
> 
> ### Use Case
> Prioritizing specific messages when possible, UCX will rely on HW configuration, for example: HW will be configured with a high-priority SL and UCX will be configured accordingly: UCX_IB_HIGH_PRIORITY_SL=<hw configured sl>
> 
> ### Implementation:
> 1. Some of the `ifaces` will have priority messages capability
> 2. Each `uct_ep` created at UCP level will have a matching high-priority `uct_ep` if iface is capable
> 3. For each request: `request_param_t` will indicate the priority level:
>    in case the selected lane / the ep supports prioritizing, the priority `uct_ep` will be used, in `ib` that ep will be configured to use the UCX_IB_HIGH_PRIORITY_SL value as sl.

Can you please specify application use case ?

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -1819,6 +1820,12 @@ typedef struct {
      */
     ucp_mem_h memh;
 
+    /**
+     * Enumerated value, indicates message priority, priority is not guaranteed 
+     * and depends on the capabilities of the selected transport,
+     * ordering across messages with different priorities are also not guaranteed.
```
Comment (shamisp) [inline src/ucp/api/ucp.h:1826]: Does it mean that order for messages with the same priorities is guaranteed ? IMHO it becomes really confusing really quickly.

Comment (shamisp) [reply src/ucp/api/ucp.h:1826]: how does it co-exist with TAGed message semantics ?

Comment (roiedanino) [reply src/ucp/api/ucp.h:1826]: > Can you please specify application use case ?
Realm with UCX backend, Realm is required to support priority messages, a kind of data-path / control-path situation in which we don't want the control messages to wait behind large data messages. They currently have a workaround (creating another worker) which shows better results.

> Does it mean that order for messages with the same priorities is guaranteed ? IMHO it becomes really confusing really quickly.
Yes, order is guaranteed as long as you don't play with priorities (which is kind of the point of priority messages)

> how does it co-exist with TAGed message semantics ? we somewhat define ordering semantics on top of other ordering semantics

The main goal is to add priorities to active messages, TAGed messages shouldn't be affected.

Comment (shamisp) [reply src/ucp/api/ucp.h:1826]: It is somewhat like adding "tags" to active messages.  In this case it should be explicitly defined as AM message priority.  I would not limit this to low and high. The tag can have a weight.

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -176,17 +176,17 @@ enum ucp_feature {
...
+    UCP_WORKER_PARAM_FIELD_CPU_MASK       = UCS_BIT(1), /**< Worker's CPU bitmap */
+    UCP_WORKER_PARAM_FIELD_EVENTS         = UCS_BIT(2), /**< Worker's events bitmap */
+    UCP_WORKER_PARAM_FIELD_USER_DATA      = UCS_BIT(3), /**< User data */
+    UCP_WORKER_PARAM_FIELD_EVENT_FD       = UCS_BIT(4), /**< External event file
+                                                             descriptor */
+    UCP_WORKER_PARAM_FIELD_FLAGS          = UCS_BIT(5), /**< Worker flags */
+    UCP_WORKER_PARAM_FIELD_NAME           = UCS_BIT(6), /**< Worker name */
+    UCP_WORKER_PARAM_FIELD_AM_ALIGNMENT   = UCS_BIT(7), /**< Alignment of active
+                                                             messages on the receiver */
+    UCP_WORKER_PARAM_FIELD_CLIENT_ID      = UCS_BIT(8)  /**< Client id */
```
Comment (yosefe) [inline src/ucp/api/ucp.h:189]: revert

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -430,10 +431,10 @@ enum ucp_lib_attr_field {
...
  */
 enum ucp_context_attr_field {
-    UCP_ATTR_FIELD_REQUEST_SIZE = UCS_BIT(0), /**< UCP request size */
-    UCP_ATTR_FIELD_THREAD_MODE  = UCS_BIT(1), /**< UCP context thread flag */
-    UCP_ATTR_FIELD_MEMORY_TYPES = UCS_BIT(2), /**< UCP supported memory types */
-    UCP_ATTR_FIELD_NAME         = UCS_BIT(3)  /**< UCP context name */
+    UCP_ATTR_FIELD_REQUEST_SIZE    = UCS_BIT(0), /**< UCP request size */
+    UCP_ATTR_FIELD_THREAD_MODE     = UCS_BIT(1), /**< UCP context thread flag */
+    UCP_ATTR_FIELD_MEMORY_TYPES    = UCS_BIT(2), /**< UCP supported memory types */
+    UCP_ATTR_FIELD_NAME            = UCS_BIT(3)  /**< UCP context name */
```
Comment (yosefe) [inline src/ucp/api/ucp.h:437]: revert

File: src/ucp/api/ucp.h [hunk 6]
```diff
@@ -734,12 +736,13 @@ typedef enum {
...
                                                         operation cannot be
                                                         completed immediately */
-    UCP_OP_ATTR_FLAG_MULTI_SEND     = UCS_BIT(19)  /**< optimize for bandwidth of
+    UCP_OP_ATTR_FLAG_MULTI_SEND     = UCS_BIT(19),  /**< optimize for bandwidth of
                                                         multiple in-flight operations,
                                                         rather than for the latency
                                                         of a single operation.
                                                         This flag and UCP_OP_ATTR_FLAG_FAST_CMPL
                                                         are mutually exclusive. */
+    UCP_OP_ATTR_FIELD_PRIORITY      = UCS_BIT(20)   /**< priority field */
```
Comment (yosefe) [inline src/ucp/api/ucp.h:745]: pls expand the documentation

File: src/ucp/api/ucp.h [hunk 7]
```diff
@@ -1819,6 +1827,12 @@ typedef struct {
      */
     ucp_mem_h memh;
 
+    /**
+     * Enumerated value, indicates message priority, priority is not guaranteed 
+     * and depends on the capabilities of the selected transport,
+     * ordering across messages with different priorities are also not guaranteed.
+     */
+    uint8_t priority;
```
Comment (yosefe) [inline src/ucp/api/ucp.h:1835]: maybe "int priority"?

File: src/ucp/api/ucp.h [hunk 8]
```diff
@@ -1262,6 +1265,11 @@ typedef struct ucp_worker_attr {
      * Maximum debug string size that can be filled with @ref ucp_request_query.
      */
     size_t                max_debug_string;
+
+    /**
+     * Maximum available message priority levels
+     */
+    uint8_t               max_message_priorities;
```
Comment (yosefe) [inline src/ucp/api/ucp.h:1272]: int

File: src/ucp/api/ucp_def.h [hunk 9]
```diff
@@ -761,6 +761,12 @@ typedef struct ucp_ep_params {
      */
     ucs_sock_addr_t         local_sockaddr;
 
+    /**
+     * Number of different priority levels, priority is not guarenteed,
+     * parameter must be between 1 and @ref ucp_worker_attr_t::max_message_priorities which
+     * can be acquired by @ref ucp_worker_query, default is 1.
+     */
+    uint32_t                required_num_of_priorities;
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h:769]: int num_priorities

File: src/ucp/api/ucp_def.h [hunk 10]
```diff
@@ -761,6 +761,12 @@ typedef struct ucp_ep_params {
      */
     ucs_sock_addr_t         local_sockaddr;
 
+    /**
+     * Number of different priority levels, priority is not guarenteed,
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h:765]: remove "priority is not guarenteed"

File: src/uct/api/uct.h [hunk 11]
```diff
@@ -969,7 +974,10 @@ enum uct_ep_params_field {
     UCT_EP_PARAM_FIELD_PRIV_DATA_LENGTH           = UCS_BIT(15),
 
     /** Enables @ref uct_ep_params::local_sockaddr */
-    UCT_EP_PARAM_FIELD_LOCAL_SOCKADDR             = UCS_BIT(16)
+    UCT_EP_PARAM_FIELD_LOCAL_SOCKADDR             = UCS_BIT(16),
+
+    /** Enables @ref uct_ep_params::priority */
+    UCT_EP_PARAM_FIELD_PRIORITY                   = UCS_BIT(17)
```
Comment (yosefe) [inline src/uct/api/uct.h:980]: pls remove UCT changes from this PR

File: src/ucp/api/ucp.h (hunk 6 above)
Comment (roiedanino) [reply src/ucp/api/ucp.h:745]: Other field flags in this enum:
```
UCP_OP_ATTR_FIELD_REQUEST       = UCS_BIT(0),  /**< request field */
    UCP_OP_ATTR_FIELD_CALLBACK      = UCS_BIT(1),  /**< cb field */
    UCP_OP_ATTR_FIELD_USER_DATA     = UCS_BIT(2),  /**< user_data field */
    UCP_OP_ATTR_FIELD_DATATYPE      = UCS_BIT(3),  /**< datatype field */
    UCP_OP_ATTR_FIELD_FLAGS         = UCS_BIT(4),  /**< operation-specific flags */
    UCP_OP_ATTR_FIELD_REPLY_BUFFER  = UCS_BIT(5),  /**< reply_buffer field */
    UCP_OP_ATTR_FIELD_MEMORY_TYPE   = UCS_BIT(6),  /**< memory type field */
    UCP_OP_ATTR_FIELD_RECV_INFO     = UCS_BIT(7),  /**< recv_info field */
    UCP_OP_ATTR_FIELD_MEMH          = UCS_BIT(8),  /**< memory handle field */
```
    
Would `/**< message priority field */` be fine?

File: src/ucp/api/ucp.h (hunk 7 above)
Comment (roiedanino) [reply src/ucp/api/ucp.h:1835]: Maybe `unsigned priority`?

File: src/ucp/api/ucp.h (hunk 6 above)
Comment (yosefe) [reply src/ucp/api/ucp.h:745]: Active message priority level

File: src/ucp/api/ucp.h [hunk 12]
```diff
@@ -742,7 +742,7 @@ typedef enum {
                                                         of a single operation.
                                                         This flag and UCP_OP_ATTR_FLAG_FAST_CMPL
                                                         are mutually exclusive. */
-    UCP_OP_ATTR_FIELD_PRIORITY      = UCS_BIT(20)   /**< priority field */
+    UCP_OP_ATTR_FIELD_PRIORITY      = UCS_BIT(20)   /**< active message priority field */
```
Comment (yosefe) [inline src/ucp/api/ucp.h:745]: Start with capital case

File: src/ucp/api/ucp_def.h [hunk 13]
```diff
@@ -762,11 +762,11 @@ typedef struct ucp_ep_params {
     ucs_sock_addr_t         local_sockaddr;
 
     /**
-     * Number of different priority levels, priority is not guarenteed,
+     * Number of different priority levels,
      * parameter must be between 1 and @ref ucp_worker_attr_t::max_message_priorities which
      * can be acquired by @ref ucp_worker_query, default is 1.
      */
-    uint32_t                required_num_of_priorities;
+    uint32_t                num_priorities;
```
Comment (yosefe) [inline src/ucp/api/ucp_def.h:769]: unsigned

File: src/ucp/api/ucp.h [hunk 14]
```diff
@@ -1262,6 +1265,11 @@ typedef struct ucp_worker_attr {
      * Maximum debug string size that can be filled with @ref ucp_request_query.
      */
     size_t                max_debug_string;
+
+    /**
+     * Maximum available message priority levels
+     */
+    unsigned              max_message_priorities;
```
Comment (yosefe) [inline src/ucp/api/ucp.h:1272]: rename to max_priorities to match UCP_WORKER_ATTR_FIELD_MAX_PRIORITIES

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@shamisp can you pls take a look?

**Comment by gleon99** [comment]
@shamisp  can look please?

File: src/ucp/api/ucp_def.h [hunk 15]
```diff
@@ -761,6 +761,12 @@ typedef struct ucp_ep_params {
      */
     ucs_sock_addr_t         local_sockaddr;
 
+    /**
+     * Number of different priority levels,
+     * parameter must be between 1 and @ref ucp_worker_attr_t::max_priorities which
+     * can be acquired by @ref ucp_worker_query, default is 1.
+     */
+    unsigned                num_priorities;
```
Comment (roiedanino) [inline src/ucp/api/ucp_def.h:769]: Should add that to `ucp_ep_t` as well

**Pull request closed without merging**